core/lib/dal/src/storage_web3_dal.rs

```diff
@@ -32,6 +32,46 @@ impl StorageWeb3Dal<'_, '_> {
         Ok(decompose_full_nonce(full_nonce).0)
     }
 
+    /// Returns the current *stored* nonces (i.e., w/o accounting for pending transactions) for the specified accounts.
+    pub async fn get_nonces_for_addresses(
+        &mut self,
+        addresses: &[Address],
+    ) -> sqlx::Result<HashMap<Address, Nonce>> {
+        let res = sqlx::query!(
```

Comment:
This is not a query I had in mind; it should be based on `storage_logs`, essentially encapsulating [`StorageWeb3Dal::get_values()`](https://github.com/matter-labs/zksync-era/blob/758f4877433ebdd86cc90dd592db9a7d87598586/core/lib/dal/src/storage_web3_dal.rs#L56) (you could even refactor [`MempoolFetcher`](https://github.com/matter-labs/zksync-era/blob/758f4877433ebdd86cc90dd592db9a7d87598586/core/lib/zksync_core/src/state_keeper/mempool_actor.rs#L140) to use this method as well, although it's by no means necessary – I can do this in a follow-up PR). Sorry for not being clear. One case in which your query certainly won't work is snapshot recovery; a recovered DB won't contain historic transactions (but it will contain appropriate storage values).

Relatively minor compared to the problem above, your query contradicts the doc comment; it *does* take pending non-errored transactions into account.

Reply:
In case of nonces from transactions, there are many more cases where we take into account nonces from txs table. 
Probably it deserves to be addressed all together